### PR TITLE
fix: appset gpg limitation for templated project fields

### LINKED ISSUE #19423: Upgrading to 2.12 breaks ability to use .namespace in ApplicationSet List Generator
Checklist:

**Describe the bug**

When deploying an applicationset similar to:
```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name:my-appset
  namespace: argocd
spec:
  syncPolicy:
    preserveResourcesOnDeletion: false
  generators:
    - matrix:
        generators:
          - list:
              elements:
                - cluster: prod
                  namespace: prod
                - cluster: prod
                  namespace: dev
                - cluster: dev
                  namespace: dev
          - matrix:
              generators:
                - git:
                    repoURL: https://gitlab......./.git
                    revision: main
                    directories:
                      - path: helm/environments/{{.namespace}}
                    pathParamPrefix: namespace
                - git:
                    repoURL: https://gitlab......./.git
                    revision: main
                    directories:
                      - path: helm/apps/*
                    pathParamPrefix: app
  goTemplate: true
  template:
    metadata:
      name: '{{.namespace}}-{{.app.path.basenameNormalized}}'
    spec:
      destination:
        name: '{{.cluster}}'
        namespace: '{{.namespace}}'
      project: '{{.namespace}}'
 .....
```

We get errors in the logs:

```
level=error msg="error generating params" error="error getting project : AppProject.argoproj.io \"{{.namespace}}\" not found" generator="&{0x4000b24780}"
```
and
```
level=error msg="error generating application from params" applicationset=argocd/exp-prod error="failed to get params for second generator in the matrix generator: child generator returned an error on parameter generation: error getting project : AppProject.argoproj.io \"{{.namespace}}\" not found" generator="{nil nil nil nil nil nil &MatrixGenerator{Generators:[]ApplicationSetNestedGenerator{ApplicationSetNestedGenerator{List:&ListGenerator{Elements:[]JSON{{[123 34 99 108 117 115 116 101 114 34 58 34 112 117 108 115 101 112 108 97 116 102 111 114 109 45 112 114 111 100 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 101 120 112 101 100 105 97 45 112 114 111 100 117 99 116 105 111 110 34 125]},{[123 34 99 108 117 115 116 101 114 34 58 34 112 117 108 115 101 112 108 97 116 102 111 114 109 45 112 114 111 100 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 99 119 103 45 112 114 111 100 117 99 116 105 111 110 34 125]},{[123 34 99 108 117 115 116 101 114 34 58 34 119 114 45 114 105 109 115 45 112 114 111 100 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 119 114 45 114 105 109 115 45 112 114 111 100 34 125]},{[123 34 99 108 117 115 116 101 114 34 58 34 103 108 111 98 97 108 45 112 108 97 116 102 111 114 109 45 112 114 111 100 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 103 108 111 98 97 108 45 112 108 97 116 102 111 114 109 45 112 114 111 100 34 125]},{[123 34 99 108 117 115 116 101 114 34 58 34 112 117 108 115 101 112 108 97 116 102 111 114 109 45 112 114 111 100 34 44 34 110 97 109 101 115 112 97 99 101 34 58 34 112 117 108 115 101 112 108 97 116 102 111 114 109 45 112 114 111 100 117 99 116 105 111 110 34 125]},},Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},ElementsYaml:,},Clusters:nil,Git:nil,SCMProvider:nil,ClusterDecisionResource:nil,PullRequest:nil,Matrix:nil,Merge:nil,Selector:nil,Plugin:nil,},ApplicationSetNestedGenerator{List:nil,Clusters:nil,Git:&GitGenerator{RepoURL:https://gitlab.devops.pulselive.io/motor-team/pdp-experience-platform.git,Directories:[]GitDirectoryGeneratorItem{GitDirectoryGeneratorItem{Path:{{.cluster}}/{{.namespace}}/*/*,Exclude:false,},},Files:[]GitFileGeneratorItem{},Revision:prod,RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},PathParamPrefix:,Values:map[string]string{},},SCMProvider:nil,ClusterDecisionResource:nil,PullRequest:nil,Matrix:nil,Merge:nil,Selector:nil,Plugin:nil,},},Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},} nil nil nil}"

level=error msg="error generating params" error="failed to get params for second generator in the matrix generator: child generator returned an error on parameter generation: error getting project : AppProject.argoproj.io \"{{.namespace}}\" not found" generator="&{map[ClusterDecisionResource:0x40009cd720 Clusters:0x40009cd6d0 Git:0x4000de5b20 List:0x74426e0 Matrix:0x4000c22a50 Merge:0x4000c22a58 Plugin:0x4000b247c0 PullRequest:0x4000dcd860 SCMProvider:0x4000dcd800]}"
```

This worked fine for us in v2.11.7.

**To Reproduce**
Use a similar applicationset to the above
**Expected behavior**

The applicationset provisions applications against a project based on the {{.namespace}} value generated by the list generator

**Version**
2.12.0

### PR DESCRIPTION
**Pull request by ishitasequeira**
Partially fix https://github.com/argoproj/argo-cd/issues/19423 and documents the limitation of using ApplicationSets Templating of project field while using signature verification with Git Generator.

cherry-pick release-2.12

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: applicationset/generators/git.go [hunk 1]
```diff
@@ -60,20 +60,31 @@ func (g *GitGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.Applic
...
 	if strings.Contains(appSet.Spec.Template.Spec.Project, "{{") {
-		project = appSetGenerator.Git.Template.Spec.Project
+		// When the project field is templated, the contents of the git repo are required to run the git generator and get the templated value,
+		// but git generator cannot be called without verifying the commit signature.
+		// In this case, we skip the signature verification.
+		if strings.Contains(appSet.Spec.Template.Spec.Project, "{{") {
+			isProjectTemplated = true
+		} else {
+			project = appSetGenerator.Git.Template.Spec.Project
+		}
```
Comment (crenshaw-dev) [inline applicationset/generators/git.go:73]: I might be reading wrong, but isn't this internal condition redundant?

File: docs/operator-manual/applicationset/Template.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ Note:
 
 - Referenced clusters must already be defined in Argo CD, for the ApplicationSet controller to use them
 - Only **one** of `name` or `server` may be specified: if both are specified, an error is returned.
+- Limitation: ApplicationSet do not support templating of project field while using signature verification.
```
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Template.md:46]: Ditto rephrasing.

File: docs/user-guide/gpg-verification.md [hunk 3]
```diff
@@ -29,6 +29,11 @@ not possible using Helm repositories.
     trust models, and it is not necessary (nor possible) to sign the public keys
     you are going to import into ArgoCD.
 
+
+!!!note Limitations
+  ApplicationSet do not support templating of project field while using signature
+  verification.
```
Comment (crenshaw-dev) [inline docs/user-guide/gpg-verification.md:35]: Ditto rephrasing.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: docs/operator-manual/applicationset/GoTemplate.md [hunk 4]
```diff
@@ -100,6 +100,16 @@ possible with Go text templates:
                   - name: throw-away
                     value: "{{end}}"
 
+- Templating the project field when using git generator with signature verification enabled
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply docs/operator-manual/applicationset/GoTemplate.md]: rephrased, let me know what you think.

File: applicationset/generators/git.go (hunk 1 above)
Comment (ishitasequeira) [reply applicationset/generators/git.go]: You are right, simplified the code logic.

**Review by crenshaw-dev — APPROVED** [review]
LGTM, thanks!

**Pull request merged**